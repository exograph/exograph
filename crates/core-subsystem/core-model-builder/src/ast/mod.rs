pub mod ast_types;
