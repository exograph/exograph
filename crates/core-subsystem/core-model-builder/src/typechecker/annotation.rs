// Copyright Exograph, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file at the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

/// Specification for an annotation.
pub struct AnnotationSpec {
    /// List of targets the annotation is allowed to be applied to.
    pub targets: &'static [AnnotationTarget],
    /// Is this annotation allowed to have no parameters?
    pub no_params: bool,
    /// Is this annotation allowed to have a single parameter?
    pub single_params: bool,
    /// List of mapped parameters if mapped parameters are allowed (`None` if not).
    pub mapped_params: Option<&'static [MappedAnnotationParamSpec]>,
}

/// Target for an annotation.
#[derive(Debug, PartialEq, Eq)]
pub enum AnnotationTarget {
    Type,
    Field,
    Argument,
    Module,
    Method,
    Interceptor,
}

/// Specification for a mapped parameter of an annotation.
pub struct MappedAnnotationParamSpec {
    /// Name of the parameter.
    pub name: &'static str,
    /// Is this parameter optional?
    pub optional: bool,
}
