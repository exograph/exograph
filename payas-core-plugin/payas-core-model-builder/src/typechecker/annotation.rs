/// Specification for an annotation.
pub struct AnnotationSpec {
    /// List of targets the annotation is allowed to be applied to.
    pub targets: &'static [AnnotationTarget],
    /// Is this annotation allowed to have no parameters?
    pub no_params: bool,
    /// Is this annotation allowed to have a single parameter?
    pub single_params: bool,
    /// List of mapped parameters if mapped parameters are allowed (`None` if not).
    pub mapped_params: Option<&'static [MappedAnnotationParamSpec]>,
}

/// Target for an annotation.
#[derive(Debug, PartialEq, Eq)]
pub enum AnnotationTarget {
    Model,
    Field,
    Argument,
    Service,
    Method,
    Interceptor,
}

/// Specification for a mapped parameter of an annotation.
pub struct MappedAnnotationParamSpec {
    /// Name of the parameter.
    pub name: &'static str,
    /// Is this parameter optional?
    pub optional: bool,
}
