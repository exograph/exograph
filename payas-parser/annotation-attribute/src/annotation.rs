use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, AttributeArgs, Fields, ItemStruct};

use crate::{is_optional, name_fn};

// TODO documentation
pub(crate) fn annotation(args: TokenStream, input: TokenStream) -> TokenStream {
    let args = parse_macro_input!(args as AttributeArgs);
    let input = parse_macro_input!(input as ItemStruct);

    let attrs = input.attrs;
    let vis = input.vis;
    let name = input.ident;
    let fields = input.fields;

    let (annot_struct, from_params_fn, pass_fn) = match fields {
        Fields::Unit => {
            let annot_struct = quote! {
                #(#attrs)*
                #vis struct #name;
            };

            // If the annotation struct has no fields, check there are no parameters
            let from_params_fn = quote! {
                fn from_params(params: TypedAnnotationParams) -> Result<Self, Vec<String>> {
                    if let TypedAnnotationParams::None = params {
                        Ok(Self {})
                    } else {
                        Err(vec!["expected no parameters".to_string()])
                    }
                }
            };

            // No args - nothing to check
            let pass_fn = quote! {
                fn pass(&mut self, params: &AstAnnotationParams, env: &MappedArena<Type>, scope: &Scope, errors: &mut Vec<codemap_diagnostic::Diagnostic>) -> bool {
                    false
                }
            };

            (annot_struct, from_params_fn, pass_fn)
        }
        Fields::Unnamed(fields) => {
            // There may only be one unnamed parameter
            if fields.unnamed.len() != 1 {
                panic!("Annotation may only have one unnamed parameter");
            }

            let field = fields.unnamed.first().unwrap();

            let annot_struct = {
                let field_type = &field.ty;
                let field_vis = &field.vis;

                quote! {
                    #(#attrs)*
                    #vis struct #name(#field_vis #field_type);
                }
            };

            let from_params_fn = {
                let body = if is_optional(&field) {
                    quote! {
                        match params {
                            TypedAnnotationParams::None => Ok(Self(None)),
                            TypedAnnotationParams::Single(expr) => Ok(Self(Some(expr))),
                            TypedAnnotationParams::Map(params) => {
                                Err(vec!["unexpected named parameters".to_string()])
                            }
                        }
                    }
                } else {
                    quote! {
                        match params {
                            TypedAnnotationParams::None => Err(vec!["expected a parameter".to_string()]),
                            TypedAnnotationParams::Single(expr) => Ok(Self(expr)),
                            TypedAnnotationParams::Map(params) => {
                                Err(vec!["unexpected named parameters".to_string()])
                            }
                        }
                    }
                };

                quote! {
                    fn from_params(params: TypedAnnotationParams) -> Result<Self, Vec<String>> {
                        #body
                    }
                }
            };

            let pass_fn = {
                let body = if is_optional(&field) {
                    quote! {
                        match &params {
                            AstAnnotationParams::None => false,
                            AstAnnotationParams::Single(expr) => {
                                expr.pass(self.0.as_mut().unwrap(), env, scope, errors)
                            }
                            AstAnnotationParams::Map(params) => panic!()
                        }
                    }
                } else {
                    quote! {
                        match &params {
                            AstAnnotationParams::None => panic!(),
                            AstAnnotationParams::Single(expr) => {
                                expr.pass(&mut self.0, env, scope, errors)
                            }
                            AstAnnotationParams::Map(params) => panic!()
                        }
                    }
                };

                quote! {
                    fn pass(&mut self, params: &AstAnnotationParams, env: &MappedArena<Type>, scope: &Scope, errors: &mut Vec<codemap_diagnostic::Diagnostic>) -> bool {
                        #body
                    }
                }
            };

            (annot_struct, from_params_fn, pass_fn)
        }
        Fields::Named(fields) => {
            let field_idents = fields
                .named
                .iter()
                .map(|f| f.ident.as_ref().unwrap())
                .collect::<Vec<_>>();

            let annot_struct = {
                let field_types = fields.named.iter().map(|f| &f.ty).collect::<Vec<_>>();
                let field_vis = fields.named.iter().map(|f| &f.vis).collect::<Vec<_>>();

                quote! {
                    #(#attrs)*
                    #vis struct #name {
                        #(#field_vis #field_idents: #field_types,)*
                    }
                }
            };

            let from_params_fn = {
                let expected_fields = field_idents
                    .iter()
                    .map(|i| i.to_string())
                    .collect::<Vec<_>>();

                let constructor_fields = field_idents
                    .iter()
                    .map(|i| {
                        let n = i.to_string();
                        quote! { #i: params[#n].clone() }
                    })
                    .collect::<Vec<_>>();

                quote! {
                    fn from_params(params: TypedAnnotationParams) -> Result<Self, Vec<String>> {
                        let mut errs = Vec::new();
                        if let TypedAnnotationParams::Map(params) = params {
                            // As we check parameters from TypedAnnotationParams, keep track of extra unused
                            // parameter
                            let mut unexpected_params = params.keys().cloned().collect::<std::collections::HashSet<_>>();

                            // For each field in the annotation struct, check if the parameter map contains
                            // the field by name
                            for expected in [#(#expected_fields),*] {
                                if params.contains_key(expected) {
                                    unexpected_params.remove(expected);
                                } else {
                                    errs.push(format!("Expected parameters {}", expected));
                                }
                            }

                            // For any unexpected parameters, push an error
                            for unexpected in unexpected_params {
                                errs.push(format!("Unexpected parameter {}", unexpected));
                            }

                            if errs.is_empty() {
                                Ok(Self {
                                    #(#constructor_fields,)*
                                })
                            } else {
                                Err(errs)
                            }
                        } else {
                            Err(vec!["expected named parameters".to_string()])
                        }
                    }
                }
            };

            let pass_fn = {
                let passes = field_idents
                    .iter()
                    .map(|i| {
                        let n = i.to_string();
                        quote! { pass = pass || params[#n].pass(&mut self.#i, env, scope, errors); }
                    })
                    .collect::<Vec<_>>();

                quote! {
                    fn pass(&mut self, params: &AstAnnotationParams, env: &MappedArena<Type>, scope: &Scope, errors: &mut Vec<codemap_diagnostic::Diagnostic>) -> bool {
                        match &params {
                            AstAnnotationParams::None => panic!(),
                            AstAnnotationParams::Single(expr) => panic!(),
                            AstAnnotationParams::Map(params) => {
                                let mut pass = false;
                                #(#passes)*
                                pass
                            }
                        }
                    }
                }
            };

            (annot_struct, from_params_fn, pass_fn)
        }
    };

    let name_fn = name_fn(&args);

    let expanded = quote! {
        #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
        #annot_struct

        impl #name {
            #from_params_fn

            #pass_fn

            #name_fn
        }
    };

    TokenStream::from(expanded)
}
