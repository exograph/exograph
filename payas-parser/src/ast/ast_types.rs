use codemap::{CodeMap, Span};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct AstSystem {
    pub models: Vec<AstModel>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct AstModel {
    pub name: String,
    pub kind: AstModelKind,
    pub fields: Vec<AstField>,
    pub annotations: Vec<AstAnnotation>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum AstModelKind {
    Persistent,
    Context,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct AstField {
    pub name: String,
    pub typ: AstFieldType,
    pub annotations: Vec<AstAnnotation>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum AstFieldType {
    Plain(
        String,
        #[serde(skip_serializing)]
        #[serde(skip_deserializing)]
        #[serde(default = "default_span")]
        Span,
    ),
    Optional(Box<AstFieldType>),
    List(Box<AstFieldType>),
}

impl AstFieldType {
    pub fn name(&self) -> String {
        match self {
            AstFieldType::Optional(underlying) | AstFieldType::List(underlying) => {
                underlying.name()
            }
            AstFieldType::Plain(base_type, _) => base_type.clone(),
        }
    }

    pub fn optional(&self) -> Self {
        match self {
            AstFieldType::Optional(_) => self.clone(),
            _ => AstFieldType::Optional(Box::new(self.clone())),
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct AstAnnotation {
    pub name: String,
    pub params: Vec<AstExpr>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum AstExpr {
    FieldSelection(FieldSelection),
    LogicalOp(LogicalOp),
    RelationalOp(RelationalOp),
    StringLiteral(
        String,
        #[serde(skip_serializing)]
        #[serde(skip_deserializing)]
        #[serde(default = "default_span")]
        Span,
    ),
    BooleanLiteral(
        bool,
        #[serde(skip_serializing)]
        #[serde(skip_deserializing)]
        #[serde(default = "default_span")]
        Span,
    ),
}

impl AstExpr {
    pub fn span(&self) -> &Span {
        match &self {
            AstExpr::FieldSelection(s) => s.span(),
            AstExpr::StringLiteral(_, s) => s,
            _ => panic!(),
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum FieldSelection {
    Single(Identifier),
    Select(
        Box<FieldSelection>,
        Identifier,
        #[serde(skip_serializing)]
        #[serde(skip_deserializing)]
        #[serde(default = "default_span")]
        Span,
    ),
}

impl FieldSelection {
    pub fn span(&self) -> &Span {
        match &self {
            FieldSelection::Select(_, _, s) => s,
            _ => panic!(),
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum LogicalOp {
    Not(
        Box<AstExpr>,
        #[serde(skip_serializing)]
        #[serde(skip_deserializing)]
        #[serde(default = "default_span")]
        Span,
    ),
    And(Box<AstExpr>, Box<AstExpr>),
    Or(Box<AstExpr>, Box<AstExpr>),
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum RelationalOp {
    Eq(Box<AstExpr>, Box<AstExpr>),
    Neq(Box<AstExpr>, Box<AstExpr>),
    Lt(Box<AstExpr>, Box<AstExpr>),
    Lte(Box<AstExpr>, Box<AstExpr>),
    Gt(Box<AstExpr>, Box<AstExpr>),
    Gte(Box<AstExpr>, Box<AstExpr>),
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Identifier(
    pub String,
    #[serde(skip_serializing)]
    #[serde(skip_deserializing)]
    #[serde(default = "default_span")]
    pub Span,
);

fn default_span() -> Span {
    let mut tmp_codemap = CodeMap::new();
    tmp_codemap
        .add_file("".to_string(), "".to_string())
        .span
        .subspan(0, 0)
}
