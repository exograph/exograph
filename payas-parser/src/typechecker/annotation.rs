use std::collections::{HashMap, HashSet};

use crate::ast::ast_types::{AstAnnotation, AstAnnotationParams, Untyped};
use codemap_diagnostic::{Diagnostic, Level, SpanLabel, SpanStyle};
use payas_model::model::mapped_arena::MappedArena;

use super::{Scope, Type, TypecheckFrom, Typed};
use crate::util;

/// Specification for an annotation.
pub struct AnnotationSpec {
    /// List of targets the annotation is allowed to be applied to.
    pub targets: &'static [AnnotationTarget],
    /// Is this annotation allowed to have no parameters?
    pub no_params: bool,
    /// Is this annotation allowed to have a single parameter?
    pub single_params: bool,
    /// List of mapped parameters if mapped parameters are allowed (`None` if not).
    pub mapped_params: Option<&'static [MappedAnnotationParamSpec]>,
}

/// Target for an annotation.
#[derive(Debug, PartialEq)]
pub enum AnnotationTarget {
    Model,
    Field,
    Argument,
    Service,
    Method,
}

/// Specification for a mapped parameter of an annotation.
pub struct MappedAnnotationParamSpec {
    /// Name of the parameter.
    pub name: &'static str,
    /// Is this parameter optional?
    pub optional: bool,
}

impl TypecheckFrom<AstAnnotation<Untyped>> for AstAnnotation<Typed> {
    fn shallow(untyped: &AstAnnotation<Untyped>) -> AstAnnotation<Typed> {
        AstAnnotation {
            name: untyped.name.clone(),
            params: AstAnnotationParams::shallow(&untyped.params),
            span: untyped.span,
        }
    }

    fn pass(
        &mut self,
        type_env: &MappedArena<Type>,
        annotation_env: &HashMap<String, AnnotationSpec>,
        scope: &Scope,
        errors: &mut Vec<Diagnostic>,
    ) -> bool {
        let spec = &annotation_env[&self.name];

        // Message for what parameters were expected in case of a type error
        let diagnostic_msg = {
            let mut expected = Vec::new();

            if spec.no_params {
                expected.push("no parameters".to_string());
            }
            if spec.single_params {
                expected.push("a single parameter".to_string());
            }
            if let Some(params) = spec.mapped_params {
                expected.push(format!(
                    "({})",
                    util::join_strings(
                        &params
                            .iter()
                            .map(|param_spec| format!(
                                "{}{}",
                                param_spec.name,
                                if param_spec.optional { "?" } else { "" }
                            ))
                            .collect::<Vec<_>>(),
                        None,
                    )
                ));
            }

            format!("expected {}", util::join_strings(&expected, Some("or")))
        };

        let base_diagnostic = Diagnostic {
            level: Level::Error,
            message: format!("Incorrect parameters for annotation `{}`", self.name),
            code: Some("A000".to_string()),
            spans: vec![SpanLabel {
                span: self.span,
                label: Some(diagnostic_msg),
                style: SpanStyle::Primary,
            }],
        };

        match &self.params {
            AstAnnotationParams::None => {
                // Given no parameters, but expected single or mapped
                if !spec.no_params {
                    errors.push(base_diagnostic);
                }
            }
            AstAnnotationParams::Single(_, span) => {
                // Given a single parameter, but expected none or mapped
                if !spec.single_params {
                    let mut diagnostic = base_diagnostic;
                    diagnostic.spans.push(SpanLabel {
                        span: *span,
                        label: Some("unexpected parameter".to_owned()),
                        style: SpanStyle::Secondary,
                    });
                    errors.push(diagnostic);
                }
            }
            AstAnnotationParams::Map(params, spans) => match spec.mapped_params {
                // Given mapped parameters, but expected none or some
                None => errors.push(base_diagnostic),

                // Check given parameters are correct
                Some(param_specs) => {
                    let mut span_labels = Vec::new();
                    let mut missing_param = false;

                    // Check for any duplicate parameters
                    for (name, spans) in spans.iter() {
                        if spans.len() > 1 {
                            let mut span_labels = vec![SpanLabel {
                                span: spans[0],
                                label: Some("previously defined here".to_owned()),
                                style: SpanStyle::Secondary,
                            }];

                            for span in &spans[1..] {
                                span_labels.push(SpanLabel {
                                    span: *span,
                                    label: Some("redefined here".to_owned()),
                                    style: SpanStyle::Primary,
                                });
                            }

                            errors.push(Diagnostic {
                                level: Level::Error,
                                message: format!("Duplicate definitions of parameter `{}`", name),
                                code: Some("A000".to_string()),
                                spans: span_labels,
                            });
                        }
                    }

                    // Keep track of extra unused parameters
                    let mut unexpected_params = params.keys().cloned().collect::<HashSet<_>>();

                    // For each field, check if it is given or if it's optional
                    for param_spec in param_specs {
                        if params.contains_key(param_spec.name) {
                            unexpected_params.remove(param_spec.name);
                        } else if !param_spec.optional {
                            missing_param = true;
                        }
                    }

                    // For any unexpected parameters, push an error
                    for unexpected in unexpected_params {
                        span_labels.push(SpanLabel {
                            span: *spans[&unexpected].first().unwrap(),
                            label: Some("unexpected parameter".to_owned()),
                            style: SpanStyle::Secondary,
                        });
                    }

                    if !span_labels.is_empty() || missing_param {
                        let mut diagnostic = base_diagnostic;
                        diagnostic.spans.append(&mut span_labels);
                        errors.push(diagnostic);
                    }
                }
            },
        }

        self.params.pass(type_env, annotation_env, scope, errors)
    }
}
