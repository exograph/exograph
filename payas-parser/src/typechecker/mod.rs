mod annotation;
mod annotation_map;
mod annotation_params;
mod expression;
mod field;
mod field_type;
mod logical_op;
mod model;
mod relational_op;
mod selection;
mod service;
mod typ;

use std::collections::HashMap;

use anyhow::{anyhow, Result};
use codemap::CodeMap;
use codemap_diagnostic::{ColorConfig, Diagnostic, Emitter};
use serde::{Deserialize, Serialize};

pub(super) use annotation_map::AnnotationMap;

pub(super) use typ::{PrimitiveType, Type};

use crate::ast::ast_types::{AstModel, AstService, NodeTypedness};
use crate::ast::ast_types::{AstSystem, Untyped};
use payas_model::model::mapped_arena::MappedArena;

use self::annotation::{AnnotationSpec, AnnotationTarget, MappedAnnotationParamSpec};

pub struct Scope {
    pub enclosing_model: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Typed;
impl NodeTypedness for Typed {
    type FieldSelection = Type;
    type RelationalOp = Type;
    type Expr = Type;
    type LogicalOp = Type;
    type Field = Type;
    type Annotations = AnnotationMap;
    type Type = bool;
}

pub trait TypecheckInto<T> {
    fn shallow(&self) -> T;
    fn pass(
        &self,
        typ: &mut T,
        type_env: &MappedArena<Type>,
        annotation_env: &HashMap<String, AnnotationSpec>,
        scope: &Scope,
        errors: &mut Vec<Diagnostic>,
    ) -> bool;
}

pub trait TypecheckFrom<T>
where
    Self: Sized,
{
    fn shallow(untyped: &T) -> Self;
    fn pass(
        &mut self,
        type_env: &MappedArena<Type>,
        annotation_env: &HashMap<String, AnnotationSpec>,
        scope: &Scope,
        errors: &mut Vec<Diagnostic>,
    ) -> bool;
}

fn populate_type_env(env: &mut MappedArena<Type>) {
    // TODO: maybe we don't need to do this manually
    env.add("Boolean", Type::Primitive(PrimitiveType::Boolean));
    env.add("Int", Type::Primitive(PrimitiveType::Int));
    env.add("Float", Type::Primitive(PrimitiveType::Float));
    env.add("Decimal", Type::Primitive(PrimitiveType::Decimal));
    env.add("String", Type::Primitive(PrimitiveType::String));
    env.add("LocalTime", Type::Primitive(PrimitiveType::LocalTime));
    env.add(
        "LocalDateTime",
        Type::Primitive(PrimitiveType::LocalDateTime),
    );
    env.add("LocalDate", Type::Primitive(PrimitiveType::LocalDate));
    env.add("Instant", Type::Primitive(PrimitiveType::Instant));
    env.add("Json", Type::Primitive(PrimitiveType::Json));

    env.add("Claytip", Type::Primitive(PrimitiveType::ClaytipInjected));

    env.add(
        "Operation",
        Type::Primitive(PrimitiveType::Interception("Operation".to_string())),
    );
}

fn populate_annotation_env(env: &mut HashMap<String, AnnotationSpec>) {
    let annotations = [
        (
            "access",
            AnnotationSpec {
                targets: &[
                    AnnotationTarget::Model,
                    AnnotationTarget::Field,
                    AnnotationTarget::Method,
                ],
                no_params: false,
                single_params: true,
                mapped_params: Some(&[
                    MappedAnnotationParamSpec {
                        name: "query",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "mutation",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "create",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "update",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "delete",
                        optional: true,
                    },
                ]),
            },
        ),
        (
            "autoincrement",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: false,
                mapped_params: None,
            },
        ),
        (
            "bits",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "column",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "dbtype",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "external",
            AnnotationSpec {
                targets: &[AnnotationTarget::Service],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "before",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "after",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "around",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "inject",
            AnnotationSpec {
                targets: &[AnnotationTarget::Argument],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "jwt",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "length",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "pk",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: false,
                mapped_params: None,
            },
        ),
        (
            "plural_name",
            AnnotationSpec {
                targets: &[AnnotationTarget::Model],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "precision",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "range",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: false,
                mapped_params: Some(&[
                    MappedAnnotationParamSpec {
                        name: "min",
                        optional: false,
                    },
                    MappedAnnotationParamSpec {
                        name: "max",
                        optional: false,
                    },
                ]),
            },
        ),
        (
            "scale",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "size",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "table",
            AnnotationSpec {
                targets: &[AnnotationTarget::Model],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
    ];

    for (name, spec) in annotations {
        env.insert(name.to_owned(), spec);
    }
}

pub fn build(ast_system: AstSystem<Untyped>, codemap: CodeMap) -> Result<MappedArena<Type>> {
    let mut ast_service_models: Vec<AstModel<Untyped>> = vec![];

    let mut types_arena: MappedArena<Type> = MappedArena::default();
    let mut annotation_env = HashMap::new();
    populate_type_env(&mut types_arena);
    populate_annotation_env(&mut annotation_env);

    let mut emitter = Emitter::stderr(ColorConfig::Always, Some(&codemap));

    for service in ast_system.services.iter() {
        ast_service_models.extend(service.models.clone());
        types_arena.add(&service.name, Type::Service(AstService::shallow(service)));
    }

    let ast_types = [ast_system.models.as_slice(), ast_service_models.as_slice()].concat();
    let ast_services = ast_system.services;

    for model in ast_types.iter() {
        types_arena.add(
            model.name.as_str(),
            Type::Composite(AstModel::shallow(model)),
        );
    }

    loop {
        let mut did_change = false;
        let init_scope = Scope {
            enclosing_model: None,
        };

        let mut errors = Vec::new();

        for model in ast_types.iter() {
            let mut typ = types_arena.get_by_key(model.name.as_str()).unwrap().clone();
            if let Type::Composite(c) = &mut typ {
                let pass_res = c.pass(&types_arena, &annotation_env, &init_scope, &mut errors);
                if pass_res {
                    *types_arena.get_by_key_mut(model.name.as_str()).unwrap() = typ;
                    did_change = true;
                }
            } else {
                panic!()
            }
        }

        for service in ast_services.iter() {
            let mut typ = types_arena
                .get_by_key(service.name.as_str())
                .unwrap()
                .clone();
            if let Type::Service(s) = &mut typ {
                let pass_res = s.pass(&types_arena, &annotation_env, &init_scope, &mut errors);
                if pass_res {
                    *types_arena.get_by_key_mut(service.name.as_str()).unwrap() = typ;
                    did_change = true;
                }
            } else {
                panic!()
            }
        }

        if !did_change {
            if !errors.is_empty() {
                emitter.emit(&errors);
                return Err(anyhow!("Could not process input clay files"));
            } else {
                return Ok(types_arena);
            }
        }
    }
}

#[cfg(test)]
pub mod test_support {
    use super::*;
    use crate::parser::*;

    pub fn build(src: &str) -> Result<MappedArena<Type>> {
        let (parsed, codemap) = parse_str(src);
        super::build(parsed, codemap)
    }

    pub fn parse_sorted(src: &str) -> Vec<(String, Type)> {
        let checked = build(src).unwrap();

        let mut entries: Vec<_> = checked
            .keys()
            .map(|key| (key.clone(), checked.get_by_key(key).unwrap().clone()))
            .collect();

        entries.sort_by_key(|pair| pair.0.to_owned());
        entries
    }
}

#[cfg(test)]
mod tests {
    use super::test_support::*;

    #[test]
    fn simple() {
        let src = r#"
        model User {
          doc: Doc @column("custom_column") @access(self.role == "role_admin" || self.role == "role_superuser" || self.doc.is_public)
          role: String
        }

        model Doc {
          is_public: Boolean
        }
        "#;

        assert_typechecking(src);
    }

    #[test]
    fn with_auth_context_use_in_field_annotation() {
        let src = r#"
        context AuthContext {
            role: String @jwt
        }
    
        model Doc {
          is_public: Boolean 
          content: String @access(AuthContext.role == "ROLE_ADMIN" || self.is_public)
        }
        "#;

        assert_typechecking(src);
    }

    #[test]
    fn with_auth_context_use_in_type_annotation() {
        let src = r#"
        context AuthContext {
            role: String @jwt
        }
    
        @access(AuthContext.role == "ROLE_ADMIN" || self.is_public)
        model Doc {
          is_public: Boolean 
          content: String 
        }
        "#;

        assert_typechecking(src);
    }

    #[test]
    fn insignificant_whitespace() {
        let typical = r#"
        @table("venues")
        model Venue {
            id: Int @column("idx") @pk
            name: String
        }
        "#;

        let with_whitespace = r#"

        @table ( "venues" )
        model    Venue   
        {
            id:   Int   @column(  "idx"  )    
            @pk 
            
            name:String

        }

        "#;

        let typical_parsed = serde_json::to_string(&parse_sorted(typical)).unwrap();
        let with_whitespace_parsed = serde_json::to_string(&parse_sorted(with_whitespace)).unwrap();
        assert_eq!(typical_parsed, with_whitespace_parsed);
    }

    #[test]
    fn unknown_annotation() {
        let src = r#"
        @asdf
        model User {
        }
        "#;

        assert_err(src);
    }

    fn duplicate_annotation() {
        let src = r#"
        @table("users")
        @table("users")
        model User {
        }
        "#;

        assert_err(src);
    }

    fn invalid_annotation_parameter_type() {
        let expected_none = r#"
        model User {
            id: Int @pk("asdf")
        }
        "#;

        let expected_single = r#"
        @table
        model User {
        }
        "#;

        let expected_map = r#"
        model User {
            id: Int @range(5)
        }
        "#;

        assert_err(expected_none);
        assert_err(expected_single);
        assert_err(expected_map);
    }

    fn duplicate_annotation_mapped_param() {
        let src = r#"
        model User {
            id: Int @range(min=5, max=10, min=3)
        }
        "#;

        assert_err(src);
    }

    fn unknown_annotation_mapped_param() {
        let src = r#"
        model User {
            id: Int @range(min=5, maxx=10)
        }
        "#;

        assert_err(src);
    }

    fn invalid_annotation_target() {
        let model = r#"
        @pk
        model User {
        }
        "#;

        let field = r#"
        model User {
            id: Int @table("asdf")
        }
        "#;

        assert_err(model);
        assert_err(field);
    }

    fn assert_typechecking(src: &str) {
        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(build(src).unwrap())
        });
    }

    fn assert_err(src: &str) {
        assert!(build(src).is_err());
    }
}
