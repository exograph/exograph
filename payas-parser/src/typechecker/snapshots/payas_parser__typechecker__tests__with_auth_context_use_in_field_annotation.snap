---
source: payas-parser/src/typechecker/mod.rs
expression: types

---
- - AuthContext
  - Composite:
      name: AuthContext
      kind: Context
      fields:
        - name: role
          ast_typ:
            Plain:
              - String
          typ:
            Reference: String
          annotations:
            annotations:
              jwt:
                name: jwt
                params: None
      annotations:
        annotations: {}
- - Boolean
  - Primitive: Boolean
- - Decimal
  - Primitive: Decimal
- - Doc
  - Composite:
      name: Doc
      kind: Persistent
      fields:
        - name: is_public
          ast_typ:
            Plain:
              - Boolean
          typ:
            Reference: Boolean
          annotations:
            annotations: {}
        - name: content
          ast_typ:
            Plain:
              - String
          typ:
            Reference: String
          annotations:
            annotations:
              access:
                name: access
                params:
                  Single:
                    - LogicalOp:
                        Or:
                          - RelationalOp:
                              Eq:
                                - FieldSelection:
                                    Select:
                                      - Single:
                                          - - AuthContext
                                          - Reference: AuthContext
                                      - - role
                                      - Reference: String
                                - StringLiteral:
                                    - ROLE_ADMIN
                                - Primitive: Boolean
                          - FieldSelection:
                              Select:
                                - Single:
                                    - - self
                                    - Reference: Doc
                                - - is_public
                                - Reference: Boolean
                          - Primitive: Boolean
      annotations:
        annotations: {}
- - Float
  - Primitive: Float
- - Instant
  - Primitive: Instant
- - Int
  - Primitive: Int
- - Json
  - Primitive: Json
- - LocalDate
  - Primitive: LocalDate
- - LocalDateTime
  - Primitive: LocalDateTime
- - LocalTime
  - Primitive: LocalTime
- - String
  - Primitive: String

